doc/superbuild/source/developers/style_guide/SourceCode.rst

```diff
@@ -58,12 +52,305 @@ namespace prefix) should be used for all public function names that are not
+============================
+
+#. Do not use language features that are not compatible with C99, C++14,
+   and MSVC v1900+ (Visual Studio 2015). Examples of such features include
+   variable-length arrays. Exceptions are allowed when interfacing with a
+   library which requires a newer standard.
+
+#. All new code added to SUNDIALS should be linted with  `clang-tidy
+   <https://clang.llvm.org/extra/clang-tidy/>`_ and formatted with `clang-format
+   <https://clang.llvm.org/docs/ClangFormat.html>`_. The ``.clang-tidy`` and
+   ``.clang-format`` files in the root of the project define our configurations for
+   the tools respectively.
+
+#. Split very long lines when it improves code readability by using ``//``.
+
+#. It may be necessary to override clang-tidy at times. This can be done with
+   the ``NOLINT`` magic comments e.g.,
+
+   .. code-block:: cpp
+
+      template<class GkoSolverType, class GkoMatrixType>
+      int SUNLinSolFree_Ginkgo(SUNLinearSolver S)
+      {
+        auto solver{static_cast<LinearSolver<GkoSolverType, GkoMatrixType>*>(S->content)};
+        delete solver; // NOLINT
+        return SUNLS_SUCCESS;
+      }
+
+      class BaseObject {
+      protected:
+        // NOLINTNEXTLINE(cppcoreguidelines-non-private-member-variables-in-classes)
+        SUNContext sunctx_{};
+      };
+
+   See the clang-tidy documentation for more details.
+
+#. Spaces not tabs.
+
+#. Comments should use proper spelling and grammar.
+
+#. Following the Google Style Guide [GoogleStyle]_, TODO comments are used to note
+   code that is "temporary, a short-term solution, or good-enough but not perfect."
+
+   A consistent TODO comment format provides an easy to search for keyword with
+   details on how to get more information. TODO comments should start with ``TODO``
+   followed by a unique identifier, enclosed in parentheses, for the person most
+   knowledgeable about the issue and a brief description of the TODO item.
+   Generally, these comments should be used sparingly and are not a substitute for
+   creating an issue or bug report. When applicable, the comment should include the
+   relevant issue or bug report number.
+
+   Examples:
+
+   .. code-block:: c
+
+     /* TODO(DJG): Update to new API in the next major release (Issue #256) */
+
+#. All SUNDIALS data structures should hold onto a ``SUNContext`` object. Exceptions 
+   are the ``SUNLogger`` and ``SUNProfiler`` classes.
+
+#. All SUNDIALS functions should return a ``SUNErrCode``. Many older functions
+   do not do this and are exceptions to the rule for backwards compatiblilty. 
+   In addition, internal helper functions may or may-not return a ``SUNErrCode``.
+
+#. All SUNDIALS functions, with the exception of some internal/private functions
+   that do not have access to a ``SUNContext``, should begin with a call to 
+   ``SUNFunctionBegin()``. This macro is used for error handling and declares a 
+   local variable access via the macro ``SUNCTX``.
```

Comment:
There were some discussions, although I think most were not on GitHub. IIRC, we determined this was acceptable. Removing it would make our code messy (SUNCheck macros would need to take a SUNContext argument) and/or require developers to always manually do what the macro is doing. Since the "hidden" variable is only in our internal code, I really dont think there is much downside.

Reply:
Agreed, that's what I remember as well. Just wanted to make sure we were all on the same page.